### PR TITLE
Advent 2017: Custom JSON unmarshaler for a GraphQL client (Dec 8)

### PR DESCRIPTION
**Pull request by shurcooL**
I had a specific story I wanted to tell. It ended up being larger than I expected. But I think/hope it's good enough despite being a large post.

This is the current state. Content is 100% complete. What's left is proof-reading and copy-editing. I will be doing that myself soon, and I welcome any additional feedback. Thanks!

P.S. I know it's late for a post scheduled for tomorrow. Sorry! It took longer than I thought.

### CODE REVIEW

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 1]
```diff
@@ -0,0 +1,558 @@
++++
+author = ["Dmitri Shuralyov"]
+date = "2017-12-08T09:15:00Z"
+title = "Custom JSON unmarshaler for a GraphQL client"
+series = ["Advent 2017"]
++++
+
+In this post, I will tell a story of how I had to build a custom JSON unmarshaler for the needs of a GraphQL client library in Go. I'll start with the history of how everything started, build motivation for why a custom JSON marshaler was truly needed, and then describe how it was implemented. This is going to be a long journey, so strap yourself in, and here we go!
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: Opening might be stronger if you focus more on the surprising aspects of the problem and less on your own journey to get there.

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+---------------------------------------
+
+In May of 2017, I set out to build the first [GraphQL client library for Go](http://graphql.org/code/#go-1). Back then, only [GraphQL server libraries](http://graphql.org/code/#go) existed in Go. My main motivation was wanting to be able to access [GitHub GraphQL API v4](https://developer.github.com/v4/) from my Go code, which had just [come out of early access](https://github.com/blog/2359-introducing-github-marketplace-and-more-tools-to-customize-your-workflow) back then. I also knew that a general-purpose GraphQL client library would be useful, enabling Go projects to access any GraphQL API. There were GraphQL clients available in other languages, and I didn't want Go users to be missing out.
+
+[GraphQL](http://graphql.org/) is a data query language for APIs, developed internally by Facebook in 2012, and made publicly available in 2015. It can be used as a replacement for, or in addition to REST APIs. It some ways, it offers significant advantages compared to REST APIs, making it an attractive option. Of course, as any newer technology, it's less mature and has some weaknesses in certain areas.
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: I would switch the first two paragraphs here.

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+
+I documented the history of my findings and design decisions made in [this issue](https://github.com/google/go-github/issues/646). I've also given a [talk](https://www.youtube.com/watch?v=mEqJbeAazow) (slides [here](https://dmitri.shuralyov.com/talks/2017/githubql/githubql.slide)) about the rationale and thinking behind many of the API and design decisions that went into the library. In this post, I want to talk about something I haven't covered before: implementing a custom JSON unmarshaler specifically for the needs of the GraphQL client library, in order to improve support for GraphQL unions.
+
+JSON Unmarshaling Task at Hand
+------------------------------
+
+Unmarshaling JSON into a structure is a very common and well understood problem. It's already implemented inside the `encoding/json` package in Go standard library. Given that JSON is such a well specified standard, why would I need to implement my own JSON unmarshaler?
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: s/why would I/why would anyone

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+```JSON
+{
+	"me": {
+		"name": "gopher",
+		"bio": "The Go gopher."
+	}
+}
+```
+
+Notice it has the same shape as the query. Taking advantage of this property turned out to be a critical factor in making the Go GraphQL client library convenient and useful.
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: I would delete the second sentence here and join the first one with the paragraph below.

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+}
+fmt.Println(query.Me.Name)
+fmt.Println(query.Me.Bio)
+
+// Output:
+// gopher
+// The Go gopher.
+```
+
+Initially, `encoding/json` was used for unmarshaling the GraphQL response into the query structure amd it worked well. But eventually, some edge cases and advanced queries were discovered, where using `encoding/json` was no longer working out.
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: I would shorten the second sentence to "But we soon found some edge cases."

Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: s/"amd it"/"and it"

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+
+Motivation for Custom JSON Unmarshaler
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: s/Motivation for Custom JSON Unmarshaler/When encoding/json Doesn't Cut it

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+			"avatarUrl": "https://golang.org/doc/gopher/run.png"
+		}
+	}
+	```
+
+	As a result, `query.Me.Photo` will not be populated, since the field is "avatarUrl" in the response, and the Go struct has a field named "Photo", which doesn't match.
+
+	This happens because `encoding/json` unmarshaler is unaware of the `graphql` struct field tags.
+
+3.	Perhaps the largest problem with using `encoding/json` came to light when looking to support the GraphQL [unions](https://facebook.github.io/graphql/October2016/#sec-Unions) feature. In GraphQL, a union is a type of object representing many objects.
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: I'd start with the largest problem TBH.

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+This motivation lead to the conclusion that for GraphQL-specific needs, a custom JSON unmarshaler is unavoidably needed.
+
+Implementing a Custom JSON Unmarshaler
+--------------------------------------
+
+Discarding a well written, thoroughly tested, battle proven JSON unmarshaler in the Go standard library and writing one from scratch is not a decision to be taken lightly. I spent considerable time looking at my options and comparing their trade-offs.
+
+Writing it from scratch would've been the last option to consider. I could've made a copy of `encoding/json` and modified it. But that would mean having to maintain a copy of `encoding/json` and keep it up to date with any upstream changes.
+
+Luckily, I found a better option. The key insight was that the process of JSON unmarshaling consists of two independent parts: parsing JSON, and populating the target struct fields with the parsed values. The JSON that GraphQL servers respond with is completely standard, specification-compliant JSON. I didn't need to make any changes there. It was only the behavior of populating target struct fields that I needed to customize.
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: I'd drop the first sentence here.

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+
+In Go 1.5, the `encoding/json` package API was expanded to expose a JSON tokenizer to the outside world. A JSON tokenizer parses JSON and returns a sequence of JSON tokens, which are higher-level and easier to work with compared to the original byte stream. I could make use of this to avoid having to parse the JSON myself.
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: s/was expanded to expose a/exposed a new/

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+Let's quickly revisit our original GraphQL unions example that wasn't working with standard `encoding/json` unmarshaler. When we replace `json.UnmarshalJSON` with `jsonutil.UnmarshalGraphQL`, the `Name` field gets populated! That's good news, it means we didn't do all that work for nothing.
+
+See proof on [playground](https://play.golang.org/p/Xfu2mqxZ5m).
+
+`jsonutil.UnmarshalGraphQL` also takes `graphql` struct field tags into account when unmarshaling, and doesn't get mislead by `json` field tags. If there are additional GraphQL-specific tweaks to unmarshaling behavior that need to be applied in the future, they'll be easy to apply. Best part is we're reusing the rigorous JSON tokenizer of `encoding/json` and its public API, so no need to deal with maintaining a fork.
+
+Conclusion
+----------
+
+If you got this far, thanks for following along with me on this journey! I hope you enjoyed it and/or learned something new.
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: I'd drop this sentence or put it at the very very end

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+
+It has been a lot of fun implementing the GraphQL client library for Go, and trying to make the best [API design decisions](https://github.com/shurcooL/githubql/issues?q=label%3A%22API+decision%22). I enjoyed using the tools that Go gives me to tackle this task. Even after using Go for 4 years, I'm still finding Go to be the absolutely most fun programming language to use, and feeling same joy I did back when I was just starting out!
+
+I think GraphQL is an exciting new technology. Its strongly typed nature is a great fit for Go. APIs that are created with it can be a pleasure to use. Keep in mind that GraphQL shines most when you're able to replace multiple REST API calls with a single carefully crafted GraphQL query. This requires high quality and completeness of the GraphQL schema, so not all GraphQL APIs are made equal.
```
Comment (kevinburke) [inline content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: s/I think GraphQL is/GraphQL is. We know who thinks it is exciting, you're writing the post!

**Comment by shurcooL** [comment]
Thanks for reviewing and making suggestions @kevinburke. I'll make appropriate changes soon.

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md (hunk 10 above)
Comment (shurcooL) [reply content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: I normally avoid sentences like this, but I let my guard down because of the holiday season, and the tone of the post felt a little more casual than usual. But oh well, I'll drop it, it doesn't add much value and  dropping it makes the post shorter.

Done.

File: content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md (hunk 1 above)
Comment (shurcooL) [reply content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md]: Point taken.

But, I'm not sure there was much surprising stuff (aside from the dual slices of values maybe), it was a pretty straightforward problem (perhaps in hindsight?). I like that the current description is very accurate and describes exactly what the post talks about. I hope it attracts people who want to learn about what it takes to unmarshal JSON and/or how GraphQL works, as boring as that might be.

In other words, I think this post largely documents the normal course of solving a problem, and the description here backs that up. I can't think of any important surprising aspects.
